Prioritize Override in Boolean Filter

epeach opened the pull request:
We got feedback that our override filter wasn't working.  https://studio.code.org/s/dance?songfilter=on. This should turn the filter on for students regardless of their age or whether they are signed in.

The boolean logic we were using was `filterOff = signedInOver13 || signedOutOver13AndNotFiltered`. Even when we were filtering (`signedOutOver13AndNoteFiltered = false`), if a student was signed in and over 13 (`signedInOver13 = true`), we would still turn the filterOff (`filterOff = true || false = true`). So this might have never worked in this case, but because we are running into issues with `signedInOver13` at the moment (https://github.com/code-dot-org/code-dot-org/pull/45233), this case is particularly relevant now.

To fix, I separated how we are tracking the override with it's own session storage token, rather than tagging it on the to the age token. Now, regardless of `user_type` or `under13`, if the override is on, we have the filter on. If the filter is off, we then go into the rest of the boolean logic that looks at `user_type` and `under13`.

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

breville commented:
First, I made a matrix to help understand what's fixed here and elsewhere:

```
                          override                 no override

signed in < 13            broken (3)               broken (2)
signed in 13+             broken (1)               works (no filter)

signed out < 13           works (filter)           works (filter)
signed out 13+            works (filter)           works (no filter)

(1) Will be fixed by this PR
(2) Will be fixed by https://github.com/code-dot-org/code-dot-org/pull/45233
(3) Interestingly will be fixed by either PR
```

epeach commented:
@breville - PTAL - thanks!

breville commented on apps/src/templates/AgeDialog.jsx:
```diff
@@ -48,7 +49,8 @@ class AgeDialog extends Component {
     // Dialog won't render
     if (queryString.parse(window.location.search).songfilter === 'on') {
-      this.setManualFilter();
+      this.props.storage.setItem(SONG_FILTER_SESSION_KEY, true);
```
Do we actually want this choice to remain in place for the session?  What if the user wants to turn the filter override off again?  I thought that relying on the URL parameter for each page load could be more appropriate.

epeach replied:
I lean towards making it hard to turn the filter off. When the student goes through the "run -> success -> continue" path and is navigated to the next page through the success dialog, the url parameter is dropped. Storing the filter in the session cookie will ensure that if a teacher wants to enforce the filter in their classroom and directs students to "studio.code.org/s/dance/lessons/1/levels/1?songfilter=on" and ask them to play through the tutorial, the filter will remain on for tutorial. I believe the '?songfilter=off' part was cut from the [original spec](https://docs.google.com/document/d/1GXo6xekcqjIQ-UGjxcGhfKxZexxxYofck8C_MLUppVM/edit#) so that students wouldn't have an easy time of turning the filter off if they've been directed to turn it on.

breville replied:
I see.  If we are changing the existing behavior it seems worth capturing that in the PR description.  I'm a bit concerned about subsequent students on shared computers getting confused, but I guess we use session storage for the signed-out age selector so this is at least consistent.

epeach replied:
I don't believe the original intended behavior has changed in regards to whether the filter override lives in a session. Previously, in `setManualFilter`, we were storing the override in the `AGE_DIALOG_SESSION_KEY`. This is also where we were storing whether a signed out user was over/under 13. This PR splits that behavior into two session keys, so we can more accurately differentiate between signedIn with/out override and signedOut with/out override.

breville commented:
I found the code for filtering rather confusing and wonder if it might be worth a small tidy-up while here.  The PR description inadvertently alludes to this complexity as it uses psuedo-variables such as `signedOutOver13AndNotFiltered` which are more descriptive than what the code actually contains.

I ended up having to make the table above to help think through each case.

Some ideas to simplify:

- Rename `setFilterStatus` to something like `getInitialFilterStatus`.  With its current name, it's not clear what its return value means.
- Consider avoiding the use of negative variable names, specifically `filterOff`, which requires inverting thinking about positive cases.
- Having a `songFilterOn` function and a `filterOff` variable is another example of having to repeatedly invert thinking.

epeach commented:
@breville - PTAL - thanks! I updated some of the variables/function name to simplify logic and added extra comments. I think it reads better now, but let me know if you agree!

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -67,27 +71,36 @@ class DanceVisualizationColumn extends React.Component {
     The filter defaults to on. If the user is over 13 (identified via account or anon dialog), filter turns off.
    */
-  setFilterStatus() {
+  initializeFilterStatus() {
```
A concern with this name is that it doesn't actually initialize anything, it just returns an initial value.  So I'd still suggest something more like `getInitialFilterStatus`.

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -8,7 +8,10 @@ import PropTypes from 'prop-types';
 import i18n from '@cdo/locale';
-import AgeDialog, {signedOutOver13} from '../templates/AgeDialog';
+import AgeDialog, {
+  signedOutUnder13 as signedOutAgeCheck,
```
`signedOutAgeCheck` seems less clear than `signedOutUnder13`.  Do we really need to alias the function here?

epeach replied:
I ended up adding the alias because otherwise there is a naming collision on line 99 `const signedOutUnder13 = signedOutAgeCheck();` I like having the const named `signedOutUnder13` there because I think it captures the logic really well since we use it in booleans with `signedInUnder13`. Aliasing the function name here felt the 'least bad' way out of that naming collision. That being said, if you have other suggestions to circumvent this, please let me know!

breville replied:
Given that it's assigned to a variable and then only used once on the next line, I think I'd just avoid the temporary assignment to a variable and have
```
return signedInUnder13 || signedOutOver13();
```

Agreed that it's nice to read two variables in the `||`, but I think the reduction in clarity in the name `signedOutAgeCheck` is a bigger downside.

If you wanted to keep the two things symmetrical, I guess you could create a `signedInUnder13()` function.

epeach replied:
If we're cool with this: `return signedInUnder13 || signedOutOver13();` I think that is easy to read and circumvent this issue perfectly. :+1:

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -40,7 +43,8 @@ const SongSelector = Radium(
           >
             {Object.keys(this.props.songData).map(
               (option, i) =>
-                (this.props.filterOff || !this.props.songData[option].pg13) && (
+                // Song should be displayed if it is not pg13 or if the filter is off.
+                (!this.props.filterOn || !this.props.songData[option].pg13) && (
```
It seems we could have a helper function which returns the appropriate list of songs, avoiding doing so much business logic in the renderer.  It would then be a great candidate to be unit tested.

epeach commented:
@breville - PTAL - thanks! Broke out that utils function and added a test, plus overall clean up.

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -26,6 +30,15 @@ const SongSelector = Radium(
+        filterOn
+      } = this.props;
+
+      let songKeys = getFilteredSongKeys(songData, filterOn);
```
nit: this can probably be a `const`.

breville commented on apps/src/dance/songs.js:
```diff
@@ -106,3 +106,15 @@ export function parseSongOptions(songManifest) {
+// Given a songManifest, returns a list of keys that represent the songs to
+// be displayed based on whether the song filter is on.
+export function getFilteredSongKeys(fullSongManifest, filterOn) {
+  let allSongKeys = Object.keys(fullSongManifest);
```
nit: can this also be a `const`?

breville commented:
It looks like the age filter UI test is failing for the last few commits.

epeach commented:
@breville - PTAL - fixed the tests - thanks!

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -67,27 +74,38 @@ class DanceVisualizationColumn extends React.Component {
-      (this.props.userType === 'student' && !this.props.under13);
-    const signedOutAge = signedOutOver13();
-    return signedInOver13 || signedOutAge;
+    // If user is signed out, query session key to determine age.
+    // Return false (no filter), if user is over 13.
+    if (userType === 'unknown') {
+      return !signedOutOver13();
+    }
```
Slightly confused by the comment & code layout here.  Presumably `userType === 'unknown'` checks whether a user is signed out, and `signedOutOver13()` queries the session key?  It could be worth shuffling a little to make these things more clear, e.g.

```
    // If user is signed out...
    if (userType === 'unknown') {
      // ...call signedOutOver13() to query session key to determine age.
      // Return false (no filter), if user is over 13.
      return !signedOutOver13();
    }
```

I think there's a second issue here: `signedOutOver13()` should probably be able to return `true` if the user is signed out and over 13, but it sounds like it isn't able to determine whether the user is signed out, so we have to wrap with this extra check first?  It seems weird to split that logic across two places, and means the function isn't able to do everything you'd expect it could.

epeach replied:
Interesting point on the second issue. After reflecting on it, I realized that the function name currently infers a lot, so I updated the name to more accurately represent what it does - not if the user has selected `over13` in the age dialog - `ageDialogSelectedOver13`.

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -67,27 +74,38 @@ class DanceVisualizationColumn extends React.Component {
+      return !signedOutOver13();
+    }
+
+    // User is signed in (student or teacher) and the filter is off.
```
By "filter is off" do we mean that the "filter override is not turned on"?

breville commented on apps/src/dance/DanceVisualizationColumn.jsx:
```diff
@@ -67,27 +74,38 @@ class DanceVisualizationColumn extends React.Component {
+    }
+
+    // User is signed in (student or teacher) and the filter is off.
+    // Return true (filterOn) if the user is under 13. Teachers assumed over13.
```
super nit: I'd replace "filterOn" with something like "filter should be turned on" since a function doesn't need to know about specific variables outside of its scope.

breville commented:
Thanks for doing the extra work to build `getFilteredSongKeys`.  The accompanying unit testing looks great.

epeach commented:
@breville - PTAL - thanks!

breville reviewed (approved):
Nice work!
